Add crossguard blog (#2042)
pulumi/docs: content/blog/announcing-crossguard-preview/index.md

---
title: "Announcing Crossguard Preview"
date: 2019-12-02
meta_desc: "Today we are announcing Pulumi CrossGuard, a Policy as Code solution, is available all Team Pro and Enterprise customers to preview."
meta_image: meta.png
authors: ["erin-krengel"]

> tags: ["crossguard", "policy-as-code", "New-Features", "Pulumi-News","Features"]

Reviewer: If we want to commit the WIP post and make changes to it later, we can set the `draft: true` field to the front matter.

Also, nit: some of the `tags` have a space after the comma, while there isn't one after `"Pulumi-News"`.

---

> Over the past few months, we have been hard at work on Pulumi CrossGuard, a Policy as Code solution. Using the `pulumi-policy` library, you can express flexible business and security rules using code. CrossGuard enables organization administrators to enforce these policies across their organization or just on specific stacks.

Reviewer: It seems like we are missing one more sentence in this paragraph, staging specific what it does. "Policy as Code" and "flexible rules" don't quite tell me what CrossGuard is if I am new to the overall concept.

What do you think of this, any better? (Feel free to ignore this suggestion if you disagree.)

"... or just on specific stacks. Allowing you to verify custom policies or enforcement of changes before they get applied to your resources."

Author: going to go w "CrossGuard allows you to verify or enforce custom policies on changes before they get applied to your resources." Wasnt really sure what was meant by "enforcement of changes "

<!--more-->

> You can run policies against any Pulumi stack, independent of the stack's language. This means you can have policies written in TypeScript and run them against a Python stack. Policies are grouped in Policy Packs and run against a stack during both `pulumi preview` and `pulumi up`. Each resource in a stack is run through the Policy Pack prior to creation, meaning you can block the creation of out-of-compliance resources.

Reviewer: Nit: "Python **-based** stack". 

Since you are introducing a new term, consider putting "Policy Packs" in italics the first time it is used.

## Why

> Using Pulumi, developers and operators are empowered to self provision their infrastructure and move quickly. One our core goals is to allow you to codify best practices and share them with your team. We often hear from organization leaders that they're concerned about how to make sure their team doesn't accidentally go off the rails. CrossGuard provides guardrails your leadership needs to allow your team to continue to operate self-sufficiently.

Reviewer: "doesn't accidentally go off the rails" is a little vague. Perhaps "... organization leaders that they are concerned about enforce security and compliance best-practices. CrossGuard provides a way to automatically enforce compliance, allowing teams to operate self-sufficiently while still ensuring organization-wide requirements are met."

## Key Features

The key features available during the CrossGuard preview are:

> * TypeScript `pulumi-policy` SDK to express custom policies

Reviewer: For each of these bullets, it seems like we should have a link. Either to the GitHub repo's or to the relevant documentation.

Author: Was able to add one to all except the last one since thats an org / update specific URL

* Run Policy Packs locally using the `--policy-pack` flag
* [AWSGuard library](https://github.com/pulumi/pulumi-awsguard) - codifies best practices for AWS
* Enforce Policy Packs across your organization or particular stacks in the Pulumi Console
* View policy errors and Policy Packs for an update in the Pulumi Console

## Pulumi CrossGuard for Everyone

Pulumi CrossGuard has something to offer everyone in your organization.

### For Organization Administrators

You want your organization to be able to move quickly and deliver business value. To do so, your team needs to be able to self-service their infrastructure. Pulumi CrossGuard gives you the peace of mind that you need to allow your team to work independently. Codify business and security policies to ensure your team meets the requirements you set forth.

> ### For Security

Reviewer: "Security" isn't a type of "Everyone". Perhaps
- For Administrators
- For Security Engineers
- For Developers and Operators

There's a lot of complexity to security and often the rest of your organization does not have all the knowledge they need to deliver software securely. Express security policies that allow you to share your knowledge and prevent issues from ending up in production.

### For Developers and Operators

There are a ton of moving pieces in our ecosystems today. It is difficult to know everything about the cloud resources we use. Pulumi CrossGuard ensure you do not fall into the trap of insecure or unreasonable defaults.

> ## Demo

Reviewer: "Demo" sounds a little contrived. Perhaps "CrossGuard in Action" for the section heading?

And rather than "in the following demo..." how about?

"""With an understanding of what CrossGuard enables, let's look at what that means in practice. The following is a policy pack definition that defines..."

---
title: "Announcing Crossguard Preview"
date: 2019-12-02
draft: true

> meta_desc: "Today we are announcing Pulumi CrossGuard, a Policy as Code solution, is available all Team Pro and Enterprise customers to preview."

Reviewer 2: ```suggestion
meta_desc: "Today we are announcing Pulumi CrossGuard, a Policy as Code solution, is available to all Team Pro and Enterprise customers to preview."
```

> meta_desc: "Today we are announcing Pulumi CrossGuard, a Policy as Code solution, is available to all Team Pro and Enterprise customers to preview."

Reviewer 3: ```suggestion
meta_desc: "Today we are announcing Pulumi CrossGuard, a Policy as Code solution that enforces custom infrastructure policies, is available for all users to preview."
```

meta_image: meta.png
authors: ["erin-krengel"]
tags: ["crossguard", "policy-as-code", "New-Features", "Pulumi-News", "Features"]
---

> Over the past few months, we have been hard at work on Pulumi CrossGuard, a Policy as Code solution. Using the `pulumi-policy` library, you can express flexible business and security rules using code. CrossGuard enables organization administrators to enforce these policies across their organization or just on specific stacks. CrossGuard allows you to verify or enforce custom policies on changes before they get applied to your resources.

Reviewer 4: ```suggestion
Over the past few months, we have been hard at work on Pulumi CrossGuard, a Policy as Code solution. Using the `pulumi-policy` library, you can express flexible business and security rules using code. CrossGuard enables organization administrators to enforce these policies across their organization or just on specific stacks. CrossGuard allows you to verify or enforce custom policies on changes before they are applied to your resources.
```

> Over the past few months, we have been hard at work on Pulumi CrossGuard, a Policy as Code solution. Using the `pulumi-policy` library, you can express flexible business and security rules using code. CrossGuard enables organization administrators to enforce these policies across their organization or just on specific stacks. CrossGuard allows you to verify or enforce custom policies on changes before they are applied to your resources.

Reviewer 2: ```suggestion
Over the past few months, we have been hard at work on Pulumi CrossGuard, a Policy as Code solution. Using the `@pulumi/policy` library, you can express flexible business and security rules using code. CrossGuard enables organization administrators to enforce these policies across their organization or just on specific stacks. CrossGuard allows you to verify or enforce custom policies on changes before they are applied to your resources.
```

> Over the past few months, we have been hard at work on Pulumi CrossGuard, a Policy as Code solution. Using the `@pulumi/policy` library, you can express flexible business and security rules using code. CrossGuard enables organization administrators to enforce these policies across their organization or just on specific stacks. CrossGuard allows you to verify or enforce custom policies on changes before they are applied to your resources.

Reviewer 3: Mainly want to avoid making this look like a JavaScript-only feature:

```suggestion
Over the past few months, we have been hard at work on Pulumi CrossGuard, a Policy as Code solution. Using CrossGuard, you can express flexible business and security rules using code. CrossGuard enables organization administrators to enforce these policies across their organization or just on specific stacks. CrossGuard allows you to verify or enforce custom policies on changes before they are applied to your resources.
```

Reviewer 3: It might be nice to add a sentence like this also:

> CrossGuard is 100% open source and available to all users of Pulumi, including the Community Edition. Advanced organization-wide policy management features are available to Team Pro and Enterprise customers.

<!--more-->

You can run policies against any Pulumi stack, independent of the stack's language. This means you can have policies written in TypeScript and run them against a Python-based stack. Policies are grouped in *Policy Packs* and run against a stack during both `pulumi preview` and `pulumi up`. Each resource in a stack is run through the Policy Pack prior to creation, meaning you can block the creation of out-of-compliance resources.

> ## Why

Reviewer 3: Might be nice here to include a handful of examples across different dimensions: security, compliance, cost, team practices (e.g. tagging), etc.

Author: Added a paragraph! let me know you what you think

> Using Pulumi, developers and operators are empowered to self provision their infrastructure and move quickly. One our core goals is to allow you to codify best practices and share them with your team. We often hear from organization leaders that they are concerned about enforce security and compliance best-practices. CrossGuard provides a way to automatically enforce compliance, allowing teams to operate self-sufficiently while still ensuring organization-wide requirements are met.

Reviewer 4: ```suggestion
Using Pulumi, developers and operators are empowered to self-provision their infrastructure and move quickly. One of our core goals is to allow you to codify best practices and share them with your team. We often hear from organization leaders that they are concerned about enforcing security and compliance best-practices. CrossGuard provides a way to automatically enforce compliance, allowing teams to operate self-sufficiently while still ensuring organization-wide requirements are met.
```

## Key Features

The key features available during the CrossGuard preview are:

> * TypeScript [pulumi-policy SDK](https://github.com/pulumi/pulumi-policy) to express custom policies

Reviewer 2: ```suggestion
* [Policy SDK](https://github.com/pulumi/pulumi-policy) to express custom policies using TypeScript or JavaScript
```

Or alternatively "@pulumi/policy package"

> * [Run Policy Packs locally](https://www.pulumi.com/docs/get-started/policy-as-code/authoring-a-policy-pack/#testing-the-policy-pack-locally) using the `--policy-pack` flag

Reviewer 2: ```suggestion
* [Run Policy Packs locally]({{< relref "/docs/get-started/policy-as-code/authoring-a-policy-pack#testing-the-policy-pack-locally" >}}) using the `--policy-pack` flag
```

> * [Run Policy Packs locally]({{< relref "/docs/get-started/policy-as-code/authoring-a-policy-pack#testing-the-policy-pack-locally" >}}) using the `--policy-pack` flag

Reviewer 3: ```suggestion
* [Run Policy Packs locally]({{< relref "/docs/get-started/policy-as-code/authoring-a-policy-pack#testing-the-policy-pack-locally" >}}) using the `--policy-pack` flag, available to all users
```

> * [AWSGuard library](https://github.com/pulumi/pulumi-awsguard) - codifies best practices for AWS

Reviewer 2: Or alternatively "@pulumi/awsguard package" (depending on what the first bullet is).

> * The [@pulumi/awsguard package](https://github.com/pulumi/pulumi-awsguard) codifies best practices for AWS

Reviewer 3: ```suggestion
* [AWSGuard](https://github.com/pulumi/pulumi-awsguard) codifies best practices for AWS
```

> * [Enforce Policy Packs](https://www.pulumi.com/docs/get-started/policy-as-code/enforcing-a-policy-pack/) across your organization or particular stacks in the Pulumi Console

Reviewer 2: ```suggestion
* [Enforce Policy Packs]({{< relref "/docs/get-started/policy-as-code/enforcing-a-policy-pack" >}}) across your organization or particular stacks in the Pulumi Console
```

> * [Enforce Policy Packs]({{< relref "/docs/get-started/policy-as-code/enforcing-a-policy-pack" >}}) across your organization or particular stacks in the Pulumi Console

Reviewer 3: ```suggestion
* [Enforce Policy Packs]({{< relref "/docs/get-started/policy-as-code/enforcing-a-policy-pack" >}}) across your organization or particular stacks in the Pulumi Console, for Team Pro and Enterprise users
```

* View policy errors and Policy Packs for an update in the Pulumi Console

## Pulumi CrossGuard for Everyone

Pulumi CrossGuard has something to offer everyone in your organization.

### For Organization Administrators

You want your organization to be able to move quickly and deliver business value. To do so, your team needs to be able to self-service their infrastructure. Pulumi CrossGuard gives you the peace of mind that you need to allow your team to work independently. Codify business and security policies to ensure your team meets the requirements you set forth.

### For Security Engineers

There's a lot of complexity to security and often the rest of your organization does not have all the knowledge they need to deliver software securely. Express security policies that allow you to share your knowledge and prevent issues from ending up in production.

### For Developers and Operators

> There are a ton of moving pieces in our ecosystems today. It is difficult to know everything about the cloud resources we use. Pulumi CrossGuard ensure you do not fall into the trap of insecure or unreasonable defaults.

Reviewer 4: ```suggestion
There are a ton of moving pieces in our ecosystems today. It is difficult to know everything about the cloud resources we use. Pulumi CrossGuard ensures you do not fall into the trap of insecure or unreasonable defaults.
```

## CrossGuard in Action

> With an understanding of what CrossGuard enables, let's look at what that means in practice. The following is a policy pack definition that prohibits creating an AWS S3 bucket that is public readable. Our Policy Pack is pretty straightforward. It uses a `validateTypedResource` helper function to filter for the resource type we care about and reports a violation if present.

Reviewer 4: ```suggestion
With an understanding of what CrossGuard enables, let's look at what that means in practice. The following is a policy pack definition that prohibits creating an AWS S3 bucket that is public readable. Our Policy Pack is not complicated. It uses a `validateTypedResource` helper function to filter for the resource type we care about and reports a violation if present.
```

```typescript
new PolicyPack("aws-typescript", {
    policies: [{
        name: "s3-no-public-read",
        description: "Prohibits setting the publicRead or publicReadWrite permission on AWS S3 buckets.",
        enforcementLevel: "mandatory",
        validateResource: validateTypedResource(aws.s3.Bucket, (bucket, args, reportViolation) => {
            if (bucket.acl === "public-read" || bucket.acl === "public-read-write") {
                reportViolation(
                    "You cannot set public-read or public-read-write on an S3 bucket. " +
                    "Read more about ACLs here: https://docs.aws.amazon.com/AmazonS3/latest/dev/acl-overview.html");
            }
        }),
    }],
});
```

We initially run `pulumi preview` for an S3 bucket that will pass the policy.

```typescript
// Create an AWS resource (S3 Bucket)
const bucket = new aws.s3.Bucket("my-bucket");
```

> We then modify the bucket to have it's access control list (ACL) set to `public-read`, which will cause our bucket to be out-of-compliance.

Reviewer 4: ```suggestion
We then modify the bucket to have its access control list (ACL) set to `public-read`, which will cause our bucket to be out-of-compliance.
```

```typescript
// Create an AWS resource (S3 Bucket)
const bucket = new aws.s3.Bucket("my-bucket", {
    acl: "public-read",
});
```

![Demo of Policy as Code](pac-demo.gif)

> While this is a simple demonstration of what you can do with Policy as Code, the framework is very flexible and allows you to write much more complex policies. Our [AWSGuard library](https://github.com/pulumi/pulumi-awsguard) is a great example of how you can use Policy as Code to write configurable policies.

Reviewer 4: ```suggestion
While this is a simple demonstration of what you can do with Policy as Code, the framework is very flexible and supports much more complex policies. Our [AWSGuard library](https://github.com/pulumi/pulumi-awsguard) is a great example of how you can use Policy as Code to write configurable policies.
```

> While this is a simple demonstration of what you can do with Policy as Code, the framework is very flexible and supports much more complex policies. Our [AWSGuard library](https://github.com/pulumi/pulumi-awsguard) is a great example of how you can use Policy as Code to write configurable policies.

Reviewer 3: I realize this is late-breaking, however we would like to message that OPA is part of our thinking here. We will be opening up the OPA repo -- similar to AWSGuard -- as an _example_ of the sort of things enabled.

So given that, what do you think about this sentence:

> Our [OPAGuard library](https://github.com/pulumi/pulumi-opaguard) demonstrates an alternative capability where CrossGuard can leverage existing ecosystems of rules, in this case any Open Policy Agent (OPA) collection of rules.

Author: I dont have access to the OPAGuard library -- can you grant me access?

## Try it Today

> Pulumi CrossGuard empowers everyone to build better, safer applications and infrastructure. Today, Pulumi CrossGuard is available to preview for all Team Pro and Enterprise organizations. Organization administrators can opt-in to the preview via the "Policies" tab in the Pulumi Console. To get started with CrossGuard today, here are some resources:

Reviewer 2: ```suggestion
Pulumi CrossGuard empowers everyone to build better, safer applications and infrastructure. Today, Pulumi CrossGuard is available to preview for all Team Pro and Enterprise organizations. Organization administrators can opt-in to the preview via the "Policies" tab in the [Pulumi Console](https://app.pulumi.com/). To get started with CrossGuard today, here are some resources:
```

> Pulumi CrossGuard empowers everyone to build better, safer applications and infrastructure. Today, Pulumi CrossGuard is available to preview for all Team Pro and Enterprise organizations. Organization administrators can opt-in to the preview via the "Policies" tab in the [Pulumi Console](https://app.pulumi.com/). To get started with CrossGuard today, here are some resources:

Reviewer 3: ```suggestion
Pulumi CrossGuard empowers everyone to build better, safer applications and infrastructure. Today, Pulumi CrossGuard is available to preview for all Pulumi users. For Team Pro and Enterprise organizations, administrators can opt-in to the preview via the "Policies" tab in the [Pulumi Console](https://app.pulumi.com/). To get started with CrossGuard today, here are some resources:
```

> * If haven't played around with Pulumi yet, here is [Pulumi's Getting Started tutorial](https://www.pulumi.com/docs/get-started/).

Reviewer 2: ```suggestion
* If haven't played around with Pulumi yet, here is [Pulumi's Getting Started tutorial]({{< relref "/docs/get-started" >}}).
```

> * [Policy as Code Getting Started tutorial](https://www.pulumi.com/docs/get-started/policy-as-code/)

Reviewer 2: ```suggestion
* [Policy as Code Getting Started tutorial]({{< relref "/docs/get-started/policy-as-code" >}})
```

> We've initially released Policy as Code with TypeScript and plan to add `pulumi-policy` SDKs for the other languages we support. We would love to hear any feedback you have! You can submit feedback via our [Contact Us form](https://www.pulumi.com/contact/) or in our [community slack](https://slack.pulumi.com/).

Reviewer 2: ```suggestion
We've initially released the capability to create policies with TypeScript or JavaScript (that work with Pulumi programs written in any supported language) and plan to add policy SDKs for other supported languages. We would love to hear any feedback you have! You can submit feedback via our [Contact Us form]({{< relref "/contact" >}}) or in our [community slack](https://slack.pulumi.com/).
```

> We've led the Policy as Code with TypeScript and plan to add `pulumi-policy` SDKs for the other languages we support. We would love to hear any feedback you have! You can submit feedback via our [Contact Us form](https://www.pulumi.com/contact/) or in our [community slack](https://slack.pulumi.com/).

Reviewer 4: ```suggestion
We've initially released Policy as Code with TypeScript and plan to add `pulumi-policy` SDKs for the other languages we support. We would love to hear any feedback you have! You can submit feedback via our [Contact Us form](https://www.pulumi.com/contact/) or in our [community slack](https://slack.pulumi.com/).
```
